dashboard/app/models/foorm/submission.rb

```diff
@@ -11,5 +11,84 @@
 #
 
 class Foorm::Submission < ActiveRecord::Base
-  belongs_to :foorm_form
+  has_one :metadata, class_name: 'Pd::WorkshopSurveyFoormSubmission', foreign_key: :foorm_submission_id
+  belongs_to :form, foreign_key: [:form_name, :form_version], primary_key: [:name, :version]
+
+  # Returns an array of hashes, each having two keys -- :question and :answer.
+  # The questions and answers are in human readable formats.
+  def formatted_answers
+    question_answer_pairs = []
+
+    parsed_answers = JSON.parse(answers)
+
+    # parsed_questions comes from the survey, and as a result will generate
+    # blank entries for questions that a respondent did not answer.
+    form.parsed_questions[:general][form.key].each do |question_id, question_details|
```

Comment:
I like that this method is self-contained, so any submission can format itself, but I'm concerned about calling parsed_questions for each submission, when we only need to call it once per form. Not sure what the best workaround for this is--we could move this logic over to the lib where the other foorm parsing logic lives, or we could have an optional parameter here so we could pass in the questions?

Reply:
Thinking about this more, and I think we will need to only use the questions as a reference for getting the human-readable question, instead of using the questions to look up answers. The reason for this is we will want to store the variables that were passed into the survey (workshop_course, workshop_subject, is_virtual, etc.) and by the nature of survey js configurations those will never be stored as questions. So we will want to put every question/answer into the csv, regardless of whether it was defined by the survey. LMK if this makes sense!

Reply:
mmm interesting. So those variables will exist in every submission for a given survey then, correct? I'm basically trying to figure out what the right set of column headers would be to put in a CSV, which you wouldn't be able to identify from either a form (doesn't have the config variables) or a submission (doesn't have questions the person didn't answer) by itself. So, I guess I'd need a form and a submission to get the right set of headers?

Reply:
could you build up the set of keys as you go? Do a first pass to get all the data, and a second pass which determines the headers based on the data and determines the set of keys? In general they will have the same variables for every submission, but I could see a case where we add a variable later on and wouldn't want to be blocked from ever including it in the csv.

Reply:
You could I guess...it seems weird to me that you could have a different set of headers for the same survey based on the set of responses that are being written to a CSV (I think this would be the case if you just generated keys from submissions, and none of the submissions had answered a given question).

Reply:
Re: another comment you made here:

> Thinking about this more, and I think we will need to only use the questions as a reference for getting the human-readable question, instead of using the questions to look up answers.

I think we do need the form itself to get human readable answers, not just questions, right? Eg, for matrices the answer being saved is here (`question_details` comes from the form):

`answer: question_details[:columns][answer]`

This is definitely tough to wrap my head around, so I may be totally on the wrong track here.

Reply:
Maybe I'm starting to see what you're saying though...maybe I just loop through the submission, and reference `form.parsed_questions` to get human readable information, instead of looping through all of the form's questions?

Reply:
yeah we will probably still need the forms to get the human readable question/answers. But yes, looping through the submission makes more sense to me, since then you won't waste time for questions that don't have answers and you can get the variables that aren't defined in the question. I agree that by and large headers should be the same, I just don't want things to break if they are different.